### PR TITLE
core: optimize C apis

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: src/Simplex/Chat/Mobile/Shared.hs [hunk 1]
```diff
@@ -1,19 +1,26 @@
...
+  pure $ BS fp $ fromIntegral len
 
 putByteString :: Ptr Word8 -> ByteString -> IO ()
-putByteString ptr bs@(PS fp offset _)  = do
-  let p = unsafeForeignPtrToPtr fp `plusPtr` offset
-  memcpy ptr p $ B.length bs
+putByteString ptr (BS fp len) =
+  withForeignPtr fp $ \p -> memcpy ptr p len
+
+newCStringFromBS :: ByteString -> IO CString
```
Comment (dpwiz) [inline src/Simplex/Chat/Mobile/Shared.hs:20]: Why not [unsafePackCString](https://hackage.haskell.org/package/bytestring-0.12.0.2/docs/Data-ByteString-Unsafe.html#v:unsafePackCString) ?

Comment (epoberezkin) [reply src/Simplex/Chat/Mobile/Shared.hs:20]: not sure why they do it like this everywhere? You mean to avoid any copying?

Comment (epoberezkin) [reply src/Simplex/Chat/Mobile/Shared.hs:20]: but ByteString can be garbage collected before CString is used.

Comment (dpwiz) [reply src/Simplex/Chat/Mobile/Shared.hs:20]: And the rest from Data.ByteString.Unsafe...

Comment (dpwiz) [reply src/Simplex/Chat/Mobile/Shared.hs:20]: I see, the code gives out `Ptr`s for outside code.

**Pull request merged**